metagraph/src/graph/alignment/aligner_labeled.hpp

```diff
@@ -0,0 +1,219 @@
+        [&](const DBGAlignment &seed) { return path_queue.get_min_path_score(seed); }
+    );
+
+    path_queue.call_alignments([&](auto&& alignment) {
+        assert(alignment.is_valid(get_graph(), &config_));
+        paths.emplace_back(std::move(alignment));
+    });
+}
+
+template <typename NodeType>
+inline void LabeledColumnExtender<NodeType>::initialize(const DBGAlignment &path) {
+    DefaultColumnExtender<NodeType>::initialize(path);
+
+    VectorMap<AnnotatedDBG::row_index, size_t> row_map;
+    row_map.reserve(path.size());
+    for (NodeType i : path) {
+        row_map[anno_graph_.graph_to_anno_index(i)]++;
+    }
+
+    auto rows = const_cast<std::vector<std::pair<AnnotatedDBG::row_index, size_t>>&&>(
+        row_map.values_container()
+    );
+
+    // pick all labels which this path could have
+    auto label_counts = anno_graph_.get_top_labels(
+        rows, anno_graph_.get_annotation().num_labels()
+    );
+
+    // encode labels and store their codes
+    target_columns_.reserve(label_counts.size());
+    target_columns_.clear();
+
+    size_t max_count = 0;
+    const auto &label_encoder = anno_graph_.get_annotation().get_label_encoder();
+    for (const auto &[label, count] : label_counts) {
+        max_count = std::max(max_count, count);
+        target_columns_.push_back(label_encoder.encode(label));
+    }
+
+    std::sort(target_columns_.begin(), target_columns_.end());
+
+    mtg::common::logger->trace("Seed has {} labels", target_columns_.size());
+
+    if (max_count != path.size()) {
+        mtg::common::logger->warn(
+            "Seed has no common labels. Maximum label multiplicity is {} / {}",
+            max_count, path.size()
+        );
+    }
+}
+
+template <typename NodeType>
+inline LabeledColumnExtender<NodeType> LabeledColumnExtender<NodeType>
+::fork_extender(std::vector<uint64_t>&& new_target_labels) const {
+    auto fork = *this;
+    fork.target_columns_ = std::move(new_target_labels);
+    return fork;
+}
+
+template <typename NodeType>
+inline std::deque<std::pair<NodeType, char>> LabeledColumnExtender<NodeType>
+::fork_extension(NodeType node,
+                 std::function<void(DBGAlignment&&, NodeType)> callback,
+                 score_t min_path_score) {
+    if (target_columns_.empty())
+        return {};
+
+    const auto &mat = anno_graph_.get_annotation().get_matrix();
+    auto base_edges = DefaultColumnExtender<NodeType>::fork_extension(
+        node, callback, min_path_score
+    );
+
+    std::deque<std::pair<NodeType, char>> edges;
+    for (const auto &edge : base_edges) {
+        const auto &[out_node, c] = edge;
+        AnnotatedDBG::row_index out_row = anno_graph_.graph_to_anno_index(out_node);
+
+        if (target_columns_.size() == 1) {
+            if (mat.get(out_row, target_columns_[0]))
+                edges.push_back(edge);
+
+        } else {
+            auto row = mat.get_row(out_row);
```

Comment:
This kind of query against the binary matrix is done with the assumption that the annotator already does some kind of result prefetching (like with RowDiff).

TODO: maybe implement a prefetching technique to work for the other BinaryMatrix types

Reply:
ah....it looks like this is only done within a `get_rows` call and not across multiple calls, so this should be implemented here to speed things up